packages/datetime/src/dateRangeInput.tsx

```diff
@@ -430,11 +490,101 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, IDat
         }
     }
 
+    private getDateRangeForCallback = (currValue?: moment.Moment, currBoundary?: DateRangeBoundary) => {
+        const otherBoundary = this.getOtherBoundary(currBoundary);
+        const otherValue = this.getStateKeysAndValuesForBoundary(otherBoundary).values.selectedValue;
+
+        const currDate = this.getDateForCallback(currValue);
+        const otherDate = this.getDateForCallback(otherValue);
+
+        return (currBoundary === DateRangeBoundary.START)
+            ? [currDate, otherDate]
+            : [otherDate, currDate];
+    }
+
+    private getDateForCallback = (momentDate: moment.Moment) => {
+        if (isMomentNull(momentDate)) {
+            return null;
+        } else if (!momentDate.isValid()) {
+            return new Date(undefined);
+        } else {
+            return fromMomentToDate(momentDate);
+        }
+    }
+
+    private getOtherBoundary = (boundary?: DateRangeBoundary) => {
+        return (boundary === DateRangeBoundary.START) ? DateRangeBoundary.END : DateRangeBoundary.START;
+    }
+
+    private doBoundaryDatesOverlap = (boundaryDate: moment.Moment, boundary: DateRangeBoundary) => {
+        const otherBoundary = this.getOtherBoundary(boundary);
+        const otherBoundaryDate = this.getStateKeysAndValuesForBoundary(otherBoundary).values.selectedValue;
+
+        // TODO: add handling for allowSingleDayRange
+        if (boundary === DateRangeBoundary.START) {
+            return boundaryDate.isSameOrAfter(otherBoundaryDate);
+        } else {
+            return boundaryDate.isSameOrBefore(otherBoundaryDate);
+        }
+    }
+
+    private isEndBoundaryThatOverlapsStartBoundary = (boundaryDate: moment.Moment, boundary: DateRangeBoundary) => {
```

Comment:
Yeah, I went with this name, because it needs to check that: 
- The `boundary` you're providing is an `END` boundary, and
- The boundary dates overlap.

Thought about something like `doesEndBoundaryOverlapStartBoundary`, but that doesn't clearly communicate that it's doing the first check above using the `boundary` parameter you provide. (In my mind, `doesEndBoundaryOverlapStartBoundary` might as well be called `doBoundariesOverlap`.)

I kind of want to just get rid of this function and move the `boundary === DateRangeBoundary.END` back inline in the two places this is used.